### ISSUE TITLE
missing match arm false positive

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#3932]
![image](https://user-images.githubusercontent.com/1711539/79016140-87039780-7b6e-11ea-8e72-b3ce2aeaae89.png)

https://github.com/rust-analyzer/rowan/blob/410909c429d107f425c30d76a2a809f830a679a5/src/cursor.rs#L497-L503

cc @JoshMcguigan

**Comment by JoshMcguigan** [comment]
Thanks for the report @matklad. I've been looking into this, but I've not been able to come up with a minimal reproduction. I suspect this is related to a type inference issue, as I don't think in `Some(child)` the `child` is really `!`?

If it is a type inference issue, it can be worked around as discussed [here](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_hir_ty/src/expr.rs#L166-L171). I did test this and it fixes this false positive in this case, but it also means the diagnostic doesn't fire when the types are actually mismatched. I still think this is probably the best option though, mismatched match arms should probably trigger a diagnostic specifically for that, rather than a missing match arm diagnostic. 

I'll submit a PR to demonstrate this.

**Comment by jonathandturner** [comment]
Another I hit:

![Screenshot from 2020-04-12 08-21-50](https://user-images.githubusercontent.com/547158/79054165-cc22e900-7c96-11ea-8dd4-6bc217b765da.png)

and here's the literal:

```rust
#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone, Hash, Deserialize, Serialize)]
pub enum Literal {
    Number(Number),
    Size(Spanned<Number>, Spanned<Unit>),
    Operator(CompareOperator),
    String(String),
    GlobPattern(String),
    ColumnPath(Vec<Member>),
    Bare(String),
}
```

**Comment by JoshMcguigan** [comment]
Thanks for the report @jonathandturner. Your case is related to support for `..` in patterns, so I'll look into expanding `hir::Pat::TupleStruct` to include that information (similar to the change made in #3905).

Sorry everyone for letting all these false positives slip through!

**Comment by JoshMcguigan** [comment]
> Your case is related to support for `..` in patterns, so I'll look into expanding `hir::Pat::TupleStruct` to include that information (similar to the change made in #3905).

Actually I think I mis-spoke here. In lowering, the [`..` is converted into a single `Pat::Wild` no matter how many fields it represents](https://github.com/rust-analyzer/rust-analyzer/blob/1a1c09ed3e3a34c0a8750f98ece9ad85595395d2/crates/ra_hir_def/src/body/lower.rs#L601). I believe this is because in lowering it isn't known how many fields it will represent. 

This causes challenges in downstream code, because it is hard to tell the difference between `Literal::Size(_)` and `Literal::Size(..)`. But at the same time, I don't think lowering yet even knows how many patterns should be in a tuple (@matklad ?), so I'm not sure this can be solved in lowering. Perhaps we really do need a `hir::Pat::DotDotPat`?

I've added test cases to #3938 to demonstrate the behavior of enum tuple structs, and I don't think this is the cause of the false positive your are seeing. My guess, at this point, is that this one is related to either our ability to resolve the type of `&self`, or some issue with how I am comparing the type of the match arms to the type of the match expression. In either case, I do think #3938 will solve that problem (potentially by turning the false positive into the possiblity of false negatives in similar cases, but this is preferred).

@jonathandturner is this example from an open source code base that you can share?

**Comment by flodiebold** [comment]
> Actually I think I mis-spoke here. In lowering, the .. is converted into a single Pat::Wild no matter how many fields it represents. I believe this is because in lowering it isn't known how many fields it will represent.

Yeah, that's wrong. Adding a `DotDotPat` is not the right solution though; rather we should store the position of the .. in a tuple struct pattern as an `Option<usize>`.

**Comment by jonathandturner** [comment]
@JoshMcguigan - this is one example: https://github.com/nushell/nushell/blob/master/crates/nu-cli/src/data/base.rs#L94-L98

**Comment by JoshMcguigan** [comment]
> ..or some issue with how I am comparing the type of the match arms to the type of the match expression..

I've confirmed the false positive you posted the screenshot of is related to this ^. The type of `&self` there is actually `&&Literal`, which is valid, but not handled correctly by [the logic I wrote to check match expression types agaist match arm patterns](https://github.com/rust-analyzer/rust-analyzer/blob/268b7987290143550461090c2c0e23371813dbec/crates/ra_hir_ty/src/expr.rs#L165-L180). 

The good news is #3938 does fix this false positive. The less good news is that we have the possibility for false negatives in cases where someone is matching against `&&MyEnum`. I suspect this would be resolved by using the type checker for comparing match arms to the match expression as suggested by @flodiebold.

@flodiebold, do you mind pointing me in the right direction on how I could get started converting this to use the type checker?

**Comment by flodiebold** [comment]
@JoshMcguigan to be honest, I don't have a very detailed plan how to do this, so let me just write down some ideas:
 - you're already using the `InferenceResult`; that's the main output of the type checker, and the idea would be that it contains everything you need.
 - if there's a type mismatch in an expression, we record that in `type_mismatches`, but we don't do that for patterns yet, it turns out:
https://github.com/rust-analyzer/rust-analyzer/blob/c388130f5ffbcbe7d3131213a24d12d02f769b87/crates/ra_hir_ty/src/infer/pat.rs#L216-L218
So we would need to add that, and then use it to check whether a pattern had the correct type.
 - for match ergonomics, we strip references from the expected type [here](https://github.com/rust-analyzer/rust-analyzer/blob/c388130f5ffbcbe7d3131213a24d12d02f769b87/crates/ra_hir_ty/src/infer/pat.rs#L99-L106). We might want to record in the inference result that this happened (and how often) so that we can choose the right type in the match check :thinking: (Also, I don't know how rustc deals with this, which would also probably be helpful.) It could be helpful to look at the [type inference tests](https://github.com/rust-analyzer/rust-analyzer/blob/c388130f5ffbcbe7d3131213a24d12d02f769b87/crates/ra_hir_ty/src/tests/patterns.rs#L315) (or add some new ones) if it's unclear what types actually get written with match ergonomics.
 - other than that, to use types to get the right constructors, the only `Ty` variant you're really interested in is [`Ty::Apply`](https://github.com/rust-analyzer/rust-analyzer/blob/c388130f5ffbcbe7d3131213a24d12d02f769b87/crates/ra_hir_ty/src/lib.rs#L296) which contains [tuples, structs, enums etc.](https://github.com/rust-analyzer/rust-analyzer/blob/c388130f5ffbcbe7d3131213a24d12d02f769b87/crates/ra_hir_ty/src/lib.rs#L79-L153)
